src/runtime/pyscope.cs

```diff
@@ -0,0 +1,535 @@
+            IntPtr ptr = Runtime.PyRun_String(
+                code, flag, _globals, _locals
+            );
+            Runtime.CheckExceptionOccurred();
+            if (ptr != Runtime.PyNone)
+            {
+                throw new PythonException();
+            }
+            Runtime.XDecref(ptr);
+        }
+
+        /// <summary>
+        /// SetVariable Method
+        /// </summary>
+        /// <remarks>
+        /// Add a new variable to the variables dict if it not exists
+        /// or update its value if the variable exists.
+        /// </remarks>
+        public void SetVariable(string name, object value)
+        {
+            IntPtr _value = Converter.ToPython(value, value?.GetType());
+            SetVariable(name, _value);
+            Runtime.XDecref(_value);
+        }
+
+        private void SetVariable(string name, IntPtr value)
+        {
+            Check();
+            using (var pyKey = new PyString(name))
+            {
+                int r = Runtime.PyObject_SetItem(variables, pyKey.obj, value);
+                if (r < 0)
+                {
+                    throw new PythonException();
+                }
+            }
+        }
+
+        /// <summary>
+        /// RemoveVariable Method
+        /// </summary>
+        /// <remarks>
+        /// Remove a variable from the variables dict.
+        /// </remarks>
+        public void RemoveVariable(string name)
+        {
+            Check();
+            using (var pyKey = new PyString(name))
+            {
+                int r = Runtime.PyObject_DelItem(variables, pyKey.obj);
+                if (r < 0)
+                {
+                    throw new PythonException();
+                }
+            }
+        }
+
+        /// <summary>
+        /// ContainsVariable Method
+        /// </summary>
+        /// <remarks>
+        /// Returns true if the variable exists in the variables dict.
+        /// </remarks>
+        public bool ContainsVariable(string name)
+        {
+            Check();
+            using (var pyKey = new PyString(name))
+            {
+                return Runtime.PyMapping_HasKey(variables, pyKey.obj) != 0;
+            }
+        }
+
+        /// <summary>
+        /// GetVariable Method
+        /// </summary>
+        /// <remarks>
+        /// Returns the value of the variable, local variable first.
+        /// If the variable is not exists, throw an Exception.
+        /// </remarks>
+        public PyObject GetVariable(string name)
+        {
+            Check();
+            using (var pyKey = new PyString(name))
+            {
+                if (Runtime.PyMapping_HasKey(variables, pyKey.obj) != 0)
+                {
+                    IntPtr op = Runtime.PyObject_GetItem(variables, pyKey.obj);
+                    if (op == IntPtr.Zero)
+                    {
+                        throw new PythonException();
+                    }
+                    if (op == Runtime.PyNone)
```

Comment:
Leaks a reference. Also, I'm not sure whether this is the right place for this conversion.